topic-operator/src/main/java/io/strimzi/operator/topic/TopicOperatorUtil.java

```diff
@@ -38,33 +37,23 @@ public class TopicOperatorUtil {
     private TopicOperatorUtil() { }
 
     /**
-     * Create a new Kafka admin client instance.
-     *
-     * @param config Topic Operator configuration.
-     * @return Kafka admin client.
-     */
-    public static Admin createKafkaAdminClient(TopicOperatorConfig config) {
-        return Admin.create(config.adminClientConfig());
-    }
-
-    /**
-     * Create a new Cruise Control client instance.
-     *
-     * @param config Topic Operator configuration.
-     * @return Cruise Control client.
+     * Map TO configuration to CC client configuration.
+     * 
+     * @param operatorCfg Topic Operator configuration.
+     * @return Cruise Control client configuration.
      */
-    public static CruiseControlClient createCruiseControlClient(TopicOperatorConfig config) {
-        return CruiseControlClient.create(
-            config.cruiseControlHostname(),
-            config.cruiseControlPort(),
-            config.cruiseControlRackEnabled(),
-            config.cruiseControlSslEnabled(),
-            config.cruiseControlSslEnabled() ? getFileContent(config.cruiseControlCrtFilePath()) : null,
-            config.cruiseControlAuthEnabled(),
-            config.cruiseControlAuthEnabled() 
-                ? new String(getFileContent(config.cruiseControlApiUserPath()), StandardCharsets.UTF_8) : null,
-            config.cruiseControlAuthEnabled() 
-                ? new String(getFileContent(config.cruiseControlApiPassPath()), StandardCharsets.UTF_8) : null
+    public static CruiseControlClient.Config mapToCcClientCfg(TopicOperatorConfig operatorCfg) {
+        return new CruiseControlClient.Config(
+            operatorCfg.cruiseControlHostname(),
+            operatorCfg.cruiseControlPort(),
+            operatorCfg.cruiseControlRackEnabled(),
+            operatorCfg.cruiseControlSslEnabled(),
+            operatorCfg.cruiseControlSslEnabled() ? getFileContent(operatorCfg.cruiseControlCrtFilePath()) : null,
+            operatorCfg.cruiseControlAuthEnabled(),
+            operatorCfg.cruiseControlAuthEnabled()
+                ? new String(getFileContent(operatorCfg.cruiseControlApiUserPath()), StandardCharsets.UTF_8) : null,
+            operatorCfg.cruiseControlAuthEnabled()
+                ? new String(getFileContent(operatorCfg.cruiseControlApiPassPath()), StandardCharsets.UTF_8) : null
```

Comment:
What exactly is the point of this change? It seems like you didn't really improved anything just made it more complicated.

The methods for the Kubernetes client or Admin client are just forwarding the call somewhere else without doing anything. But this seems to have some logic and save some code / improve readability. So maybe this should be kept as it was?

Reply:
It's a simple config mapping. The improvement here is in the CC client API. In terms of util method, it doesn't change much. Let me also add a test.

Reply:
We trade one method for two methods and a class. That does not seem like a simplification to me.

Reply:
The CC client "create" methods now takes a config object as input, like the other two clients (Kube and Admin). This is the improvement.

Reply:
Sorry, but ... for me it just creates yet another unnecessary class (record) that is essentially used in a single place.

Reply:
For me it's necessary to have a better CC client API. I'm moving this to config class, where we have similar logic for Kafka admin client, and adding tests. Let's see after that.

Reply:
For me, it is completely unnecessary and it makes the code just more complicated instead of simplifying it and _removing unnecessary methods_.

Reply:
You have a single method calling a constructor with a bunch of parameters. You do not need to insert another class/record in between them to encapsulate these parameters. It does not add any value. It does not help with the readability or reusability. It just adds more _empty_ lines of code without helping with anything.

Reply:
Ok, I reworked the code adding CC client config validation that was missing, and TO config method with additional validations (this is the same logic that we have for the admin client). I also added new TO config tests.

Reply:
Sorry, but I think the main issue I have with the code is still the same. I do not think we need an extra configuration class (`CruiseControlClient.Config`) just to stand between the constructor and the single place where it is called. It just makes it harder to read and consuems memory. It seems to me like the original `createCruiseControlClient` was mostly correct in this case. The validation can be either added there or to the constructor.

Reply:
Like the Kafka admin and Kubernetes clients, the CC client was designed as a general purpose client that can potentially be reused. As part of this change, I'm also adding CC client input validation and making its creation consistent with the other clients. This is why the extra class and logic are needed.

Reply:
But it is not a general-purpose client. It is not used by anything else or even shared with anything other than Topic OPerator itself. If it one day becomes a general-purpose client, then it maybe _might make a little bit of sense_. But we are not there and I do not think there is any proposal to get there either. So maybe we should not update the client to that just yet.

Reply:
Sorry, but I don't understand what is the real problem in creating a simple new POJO that groups together all CC configuration parameters. The increase in complexity is negligible. Overall, I think it is a design improvement.

Reply:
Unnecessary complexity, unnecessary code, increased memory usage. These are the clear negatives. What are the positives to put against it? I do not see any 

TBH, the current iteration seems to be especially bad. Even if I accepted the premise of the shared client:
* The validation is only in the TO method, so not ready for the client to be _shared_ anyway. If it is expected to be a shared class, it should be in the constructor so that everyone goes through the validation.
* You still have to always pass all 8 parameters as before, you just pass it to a different class/record and then pass that to the constructor. That just creates a completely unnecessary middleman.

You argue that the Kubernetes Client does it that way as well. But:
* Kubernetes Client is a library expected to be used by many projects and designed by someone else. It is not an internal class used by Strimzi and how it is used was not our decision. So it is hard to argue about it as a pattern for internal Strimzi classes. Yes, if we one day decide to create a generally usable Cruise Control client open to the public (i.e. not just shared between CO and TO, but designed as a client for a general public), it might be a good inspiration - but we are not there yet.
* There are also clear differences. The config is optional, and it uses a builder to build the config if needed so you actually need to provide only the values you need to provide. So the middleman it creates is significantly more sophisticated.

The Kafka Admin client is another example you use. Not sure if you mean the Strimzi `AdminClientProvider` or the actual Kafka client, but ...
* For the actual Kafka Admin client, similar things that apply to the Kubernetes Client apply. It is not an internal Strimzi class designed by us. Also, it does not use a config class as the parameter that takes every single Kafka Admin client option as a parameter. It uses an opaque Map or Properties class that you can populate with the configuration options you want.
* The `AdminClientProvider` is a better example -> an internal class of Strimzi. In this case, it is actually shared between various Strimzi projects and not used just by one of them. But it does not really use a single config class to pass all parameters to it. It uses dedicated fields passed to the methods each for the different ocnfiguration option (e.g. hostname, trusted certs, user certs, config). So it does not seem to use the same pattern you are trying to use here.

Reply:
> The validation is only in the TO method

Validation is in `CruiseControlClientImpl.createInternal`.

> You still have to always pass all 8 parameters as before, you just pass it to a different class/record and then pass that to the constructor. That just creates a completely unnecessary middleman.

This additional class/record has the benefit of grouping some logically related TO configurations (CC ones), and pass them around as a single entity.